File: web-sdk/src/domain/models/chat-message/chat-message.ts
// Copyright (c) Yalochat, Inc. All rights reserved.

import type { Product } from '../product/product';

export const MessageRoles = ['USER', 'AGENT'] as const;
export type MessageRole = (typeof MessageRoles)[number];

export const MessageButtonTypes = ['reply', 'postback', 'link'] as const;
export type MessageButtonType = (typeof MessageButtonTypes)[number];

export type MessageButton = {
  readonly text: string;
  readonly type: MessageButtonType;
  readonly url?: string;
};

export const MessageTypes = [
  'text',
  'image',
  'voice',
  'product',
  'productCarousel',
  'promotion',
  'video',
  'attachment',
  'unknown',
] as const;
export type MessageType = (typeof MessageTypes)[number];

export const MessageStatuses = [
  'DELIVERED',
  'READ',
  'ERROR',
  'SENT',
  'IN_PROGRESS',
] as const;
export type MessageStatus = (typeof MessageStatuses)[number];

export class ChatMessage {
  readonly id?: number;
  readonly wiId?: string;
  readonly role: MessageRole;
  readonly content: string;
  readonly type: MessageType;
  readonly status: MessageStatus;
  readonly timestamp: Date;
  readonly fileName?: string;
  readonly amplitudes?: number[];
  readonly duration?: number;
  readonly byteCount?: number;
  readonly mediaType?: string;
  readonly blob?: Blob;
  readonly header?: string;
  readonly footer?: string;
  readonly buttons: MessageButton[];
  readonly products: Product[];
  readonly expand: boolean;

  constructor(params: {
    role: MessageRole;
    type: MessageType;
    timestamp: Date;
    id?: number;
    wiId?: string;
    content?: string;
    status?: MessageStatus;
    fileName?: string;
    amplitudes?: number[];
    duration?: number;
    byteCount?: number;
    mediaType?: string;
    blob?: Blob;
    header?: string;
    footer?: string;
    buttons?: MessageButton[];
    products?: Product[];
    expand?: boolean;
  }) {
    this.id = params.id;
    this.wiId = params.wiId;
    this.role = params.role;
    this.type = params.type;
    this.timestamp = params.timestamp;
    this.content = params.content ?? '';
    this.status = params.status ?? 'IN_PROGRESS';
    this.fileName = params.fileName;
    this.amplitudes = params.amplitudes;
    this.duration = params.duration;
    this.byteCount = params.byteCount;
    this.mediaType = params.mediaType;
    this.blob = params.blob;
    this.header = params.header;
    this.footer = params.footer;
    this.buttons = params.buttons ?? [];
    this.products = params.products ?? [];
    this.expand = params.expand ?? false;
  }

  static text(params: {
    role: MessageRole;
    timestamp: Date;
    content: string;
    id?: number;
    wiId?: string;
    status?: MessageStatus;
    header?: string;
    footer?: string;
    buttons?: MessageButton[];
  }): ChatMessage {
    return new ChatMessage({ ...params, type: 'text' });
  }

  static voice(params: {
    role: MessageRole;
    timestamp: Date;
    fileName: string;
    amplitudes: number[];
    duration: number;
    id?: number;
    wiId?: string;
    status?: MessageStatus;
    byteCount?: number;
    mediaType?: string;
    blob?: Blob;
    header?: string;
    footer?: string;
    buttons?: MessageButton[];
  }): ChatMessage {
    return new ChatMessage({ ...params, type: 'voice' });
  }

  static image(params: {
    role: MessageRole;
    timestamp: Date;
    fileName: string;
    id?: number;
    wiId?: string;
    status?: MessageStatus;
    content?: string;
    byteCount?: number;
    mediaType?: string;
    blob?: Blob;
    header?: string;
    footer?: string;
    buttons?: MessageButton[];
  }): ChatMessage {
    return new ChatMessage({ ...params, type: 'image' });
  }

  static video(params: {
    role: MessageRole;
    timestamp: Date;
    fileName: string;
    duration: number;
    id?: number;
    wiId?: string;
    status?: MessageStatus;
    content?: string;
    byteCount?: number;
    mediaType?: string;
    blob?: Blob;
    header?: string;
    footer?: string;
    buttons?: MessageButton[];
  }): ChatMessage {
    return new ChatMessage({ ...params, type: 'video' });
  }

  static attachment(params: {
    role: MessageRole;
    timestamp: Date;
    fileName: string;
    id?: number;
    wiId?: string;
    status?: MessageStatus;
    content?: string;
    byteCount?: number;
    mediaType?: string;
    blob?: Blob;
    header?: string;
    footer?: string;
    buttons?: MessageButton[];
  }): ChatMessage {
    return new ChatMessage({ ...params, type: 'attachment' });
  }

  static product(params: {
    role: MessageRole;
    timestamp: Date;
    products: Product[];
    id?: number;
    wiId?: string;
    status?: MessageStatus;
    expand?: boolean;
  }): ChatMessage {
    return new ChatMessage({ ...params, type: 'product' });
  }

  static carousel(params: {
    role: MessageRole;
    timestamp: Date;
    products: Product[];
    id?: number;
    wiId?: string;
    status?: MessageStatus;
    expand?: boolean;
  }): ChatMessage {
    return new ChatMessage({ ...params, type: 'productCarousel' });
  }
}

File: flutter-sdk/lib/src/domain/models/chat_message/chat_message.dart
// Copyright (c) Yalochat, Inc. All rights reserved.

import 'package:yalo_chat_flutter_sdk/domain/models/product/product.dart';
import 'package:equatable/equatable.dart';

import 'button.dart';

enum MessageRole {
  user('USER'),
  assistant('AGENT');

  final String role;

  const MessageRole(this.role);
}

enum MessageType {
  text('text'),
  image('image'),
  voice('voice'),
  video('video'),
  product('product'),
  productCarousel('productCarousel'),
  promotion('promotion'),

  unknown('unknown');

  final String type;
  const MessageType(this.type);
}

enum MessageStatus {
  delivered('DELIVERED'),
  read('READ'),
  error('ERROR'),
  sent('SENT'),
  inProgress('IN_PROGRESS');

  final String status;
  const MessageStatus(this.status);
}

// A class that represents a chat message in the chat
class ChatMessage extends Equatable {
  final int? id;
  final String? wiId;
  final MessageRole role;
  final String content;
  final MessageType type;
  final MessageStatus status;
  final DateTime timestamp;

  // Attached file name to resource of the message
  // for images the image file path, and for audios the audio path
  final String? fileName;

  // Audio amplitudes preview
  final List<double>? amplitudes;
  // Audio duration in ms
  final int? duration;

  // File size in bytes (for image and voice messages)
  final int? byteCount;

  // MIME type of the media file (for image and voice messages)
  final String? mediaType;

  // Products linked to the chat message
  final List<Product> products;

  // Used to indicate if the expanded version of a message should be displayed
  // this should not be persisted in DB
  final bool expand;

  // Optional header text shown above the message content
  final String? header;

  // Optional footer text shown below the message content
  final String? footer;

  // Buttons attached to the message, matching the proto Button schema.
  final List<Button> buttons;

  const ChatMessage({
    this.id,
    this.wiId,
    required this.role,
    required this.type,
    required this.timestamp,
    this.content = '',
    this.status = MessageStatus.inProgress,
    this.fileName,
    this.amplitudes,
    this.duration,
    this.byteCount,
    this.mediaType,
    this.products = const [],
    this.expand = false,
    this.header,
    this.footer,
    this.buttons = const [],
  });

  const ChatMessage.text({
    this.id,
    this.wiId,
    required this.role,
    required this.timestamp,
    this.status = MessageStatus.inProgress,
    required this.content,
    this.header,
    this.footer,
    this.buttons = const [],
  }) : type = MessageType.text,
       amplitudes = null,
       fileName = null,
       duration = null,
       byteCount = null,
       mediaType = null,
       products = const [],
       expand = false;

  const ChatMessage.voice({
    this.id,
    this.wiId,
    required this.role,
    required this.timestamp,
    this.status = MessageStatus.inProgress,
    required this.fileName,
    required this.amplitudes,
    required this.duration,
    required this.byteCount,
    required this.mediaType,
    this.header,
    this.footer,
    this.buttons = const [],
  }) : type = MessageType.voice,
       products = const [],
       content = '',
       expand = false;

  const ChatMessage.video({
    this.id,
    this.wiId,
    required this.role,
    required this.timestamp,
    required this.fileName,
    required this.duration,
    required this.byteCount,
    required this.mediaType,
    this.status = MessageStatus.inProgress,
    this.content = '',
    this.header,
    this.footer,
    this.buttons = const [],
  }) : type = MessageType.video,
       amplitudes = null,
       products = const [],
       expand = false;

  const ChatMessage.image({
    this.id,
    this.wiId,
    required this.role,
    required this.timestamp,
    required this.fileName,
    this.status = MessageStatus.inProgress,
    this.content = '',
    required this.byteCount,
    required this.mediaType,
    this.header,
    this.footer,
    this.buttons = const [],
  }) : type = MessageType.image,
       amplitudes = null,
       duration = null,
       products = const [],
       expand = false;

  const ChatMessage.product({
    this.id,
    this.wiId,
    required this.role,
    required this.timestamp,
    this.status = MessageStatus.inProgress,
    this.products = const [],
    this.expand = false,
  }) : type = MessageType.product,
       content = '',
       fileName = '',
       amplitudes = null,
       duration = null,
       byteCount = null,
       mediaType = null,
       header = null,
       footer = null,
       buttons = const [];

  const ChatMessage.carousel({
    this.id,
    this.wiId,
    required this.role,
    required this.timestamp,
    this.status = MessageStatus.inProgress,
    this.products = const [],
    this.expand = false,
  }) : type = MessageType.productCarousel,
       content = '',
       fileName = '',
       amplitudes = null,
       duration = null,
       byteCount = null,
       mediaType = null,
       header = null,
       footer = null,
       buttons = const [];

  // Creates a copy of a chat message
  ChatMessage copyWith({
    int? id,
    String? wiId,
    MessageRole? role,
    String? content,
    MessageType? type,
    MessageStatus? status,
    String? fileName,
    List<double>? amplitudes,
    int? duration,
    int? byteCount,
    String? mediaType,
    List<Product>? products,
    bool? expand,
    DateTime? timestamp,
    String? header,
    String? footer,
    List<Button>? buttons,
  }) {
    return ChatMessage(
      id: id ?? this.id,
      wiId: wiId ?? this.wiId,
      role: role ?? this.role,
      content: content ?? this.content,
      type: type ?? this.type,
      status: status ?? this.status,
      fileName: fileName ?? this.fileName,
      amplitudes: amplitudes ?? this.amplitudes,
      duration: duration ?? this.duration,
      byteCount: byteCount ?? this.byteCount,
      mediaType: mediaType ?? this.mediaType,
      products: products ?? this.products,
      expand: expand ?? this.expand,
      timestamp: timestamp ?? this.timestamp,
      header: header ?? this.header,
      footer: footer ?? this.footer,
      buttons: buttons ?? this.buttons,
    );
  }

  // Equatable props array
  @override
  List<Object?> get props => [
    id,
    wiId,
    role,
    content,
    type,
    status,
    fileName,
    amplitudes,
    duration,
    byteCount,
    mediaType,
    products,
    expand,
    timestamp,
    header,
    footer,
    buttons,
  ];
}
